### ISSUE TITLE
Check `mikz/miser`

### ISSUE THREAD

**Issue by mikz** [bankscrap/bankscrap#13]
Hi!
I would consider joining effort by porting my bank scrapers from https://github.com/mikz/miser

But I took very different approach, using real browser and capybara instead of doing all that hand. 

Would you consider it as a optional dependency?

**Comment by javiercr** [comment]
Wow, that's awesome! We weren't aware about your work with miser.

We prefer finding real APIs to inreact with, but we're ok with web scrappers too. In fact the implementation we have for Bankinter (done by @ismaGNU) uses Nokogiri for scrapping their site.

I think we should definitely join efforts.

Right now, and after trying multiple options, we're using [mechanize](https://github.com/sparklemotion/mechanize) for HTTP requests, which has an API similar to capybara. Before porting your scrapers, do you think could it be possible to rewrite them with mechanize instead of capybara? To avoid adding additional dependencies.

Also, have you tried to find APIs for banc_sabadell and evo_banco? Check out the [slides of the talk we did in Madrid.rb](http://www.slideshare.net/Diacode/hacking-your-bank-with-ruby-and-reverse-engineering-madridrb) which explains how to do it (video will be out soon):


I'm really excited to see this happening :clap:!

**Comment by mikz** [comment]
@javiercr yeah I tried using mechanize, but there is just too much javascript that I didn't want to bother with it. 

What about having the banks as separate gems which can have own dependencies?

I'm all for discovering real APIs and I already tried. Not sure which one of them, but I think eve_banco was really weird and I could not get anything, but sabadell was pretty fine. 

Also I have this prototype branch https://github.com/mikz/miser/pull/2 that is a website, but encrypts your credentials by master secret, so you can host it on heroku. But that all can be extension to bank scrap.

Will check the APIs and get back to you in few days.

**Comment by ismaGNU** [comment]
Hi, I am glad to hear about miser, It is music for our ears.

Like @javiercr says, could you try to avoid using capybara? I have tried capybara before but It has strong dependencies. Actually, you need a xserver and real browser to begin with, that was my reason to move on to http requests and html parsers. 

I understand if you don´t want to migrate, It is tedious. But, if you want, I could help you with the migrations, I think I can get an evo bank account to try. 

IMO, splitting banks into gems its not a good idea. The goal of this gem is to unify. It´s a same that dependencies could not be optional.

Anyway, thanks for taking the time to consider the migration. Let us know what are your improvement checking the APIs.

**Comment by javiercr** [comment]
@mikz is that JS all mandatory? @ismaGNU had to run some JS and used execjs gem for that, but I understand that in some case we could need a full browser.

Splitting into multiple gems is an option (as long as we keep a main gem that would be a dependency for all the individual gems), but we went for the single gem approach following the model used by [ActiveMerchant](https://github.com/Shopify/active_merchant), which I think it makes easier for newcomers to write new adapters for other banks. Our idea was to add a script that could generate a boilerplate adapter, with the proper class and the methods that need to be implemented in order to be compatible with the bank_scrap's API. @ferblape @raulmarcosl what do you think about this?

Regarding the site app, @raulmarcosl was working in another gem that would be something very similar, a web app for display bank_scrap data in a nice format (with charts and all of that).

**Comment by mikz** [comment]
@javiercr No idea if it is mandatory. Probably it could be replaced by execjs. But I just found easier to run capybara. 

The drivers are really simple: https://github.com/mikz/miser/blob/master/lib/miser/driver/banc_sabadell.rb and https://github.com/mikz/miser/blob/master/lib/miser/driver/evo_banco.rb.

After trying firefox, chrome, capybara-webkit the easiest one is poltergeist which is wrapper of phantomjs.
Installing phantomjs is fairly easy, but still it is a dependency.

I would not consider porting current drivers to non-browser stack, but if their API proves hackable, I'd make a driver. 

I understand the desire to unify the ecosystem and agree. However, some drivers require more things than others, for example the ING requires rmagic? What if I can't install it? Rmagic has issues on different ruby version etc etc etc. For someone it might be just easier drop statically linked phantomjs.

Having all dependencies in the main gem, is a bad idea imo. But you can still have all the drivers in one repo, and just split it to gems. Almost every major ruby project in the end splits to subgems. I understand it is more maintenance, but  when it is in one repo, it is not that bad.

Looking forward our collaboration guys!

**Comment by javiercr** [comment]
You're right about rmagick, we tried to avoid it when hacking ING but it was the easier solution for hacking the virtual keyboard (we needed image processing).

I use phantomjs in a daily basis (for Rails integration testing), and it is cool but IMHO is a bit slow until it fires up. However I must admit I haven't tried it outside Rails apps, so I'm not sure if that lack of speeds comes actually from phantomjs / poltergeist.  How much long does it takes for fetching bank transactions with your current implementation?

In other words, in my opinion for implementing bank adapters here is the order of preference:

` Real APIs > Simple Web scrappers > Web scrappers that require full browsers with JS enabled`

Do you know any other gem that uses a multi gem approach besides Rails? So we can research about their approach.

**Comment by mikz** [comment]
Sabadell takes about 10 second to get list of recent transactions. Evo looks broken now, because they changed the website.

Definitely agree on the order. I made miser because I wanted to get report how much I spent every day to my mailbox. So I went simple and dirty.

Just from top of my head: Rspec went through split in 2.x. Compass now splits cli and core. Cucumber in 2.0 is split. Capybara drivers are splitted. Github's [peek](https://github.com/peek/peek). Pry and its extensions. NewRelic's contrib. Some of them are governed by different people and some of them by the same. 

I think that properly designed driver system would allow people to have best of the both worlds. For example if the driver is good and you have a contributor that has that account, you'll pull it in if the author agrees.

**Comment by javiercr** [comment]
We'll check out those projects.

If we finally decide to stick to the one single gem approach, I'd suggest using [http://nofail.de/2010/11/gem-best-practice/](this approach) for optional dependencies:

```ruby
# require optional dependency
begin
  require 'rmagick'
rescue LoadError
  puts "ING needs rmagick to be installed, please 'gem install rmagick'"
end
```

**Comment by ferblape** [comment]
I don't have a strong opinion about splitting. I like the idea of keeping things smaller, but the examples mentioned have a lot of lines of code, much more than bank_scrap right now, so it was justified. 

Given that there are no tests right now, I'd keep things unified with the begin/rescue approach when requiring gems. But, of course, it's just an opinion.

**Comment by mikz** [comment]
Speaking of tests, I was investigating how to add tests for miser and found one decent solution.
Using VCR with https://www.relishapp.com/vcr/vcr/docs/configuration/filter-sensitive-data as a base + some rake task to do a cleanup. But for API tests it does not make much sense, so I'd settle with webmock.

 I'd definitely write tests when porting my drivers, so which tools you prefer? Minitest/rspec?

**Comment by javiercr** [comment]
Thanks for the input @ferblape!

@mikz why do you think VCR doesn't make sense for API tests? I was thinking about using it but I didn't know it had support for filtering sensitive data, I'll give it a try.

Regarding rspec vs minitest, I'd prefer rspec.

Sorry about the off topic, but me and @raulmarcosl we've been thinking about setting up a Slack channel to discuss the project. Check out this [blog post](http://rauchg.com/slackin/) and let me know if you like the idea.

**Comment by mikz** [comment]
@javiercr because you want to spec the API client, so you want to have the requests and responses in the test. You probably want to make some 'fake' calls, so for example you don't have to login every time.
Also if you want to set up expectations, you need to have some values. Those values would change if different people would run the tests (unless you want to share an account with them).

Regardin Slack, I welcome the idea. It has IRC gateway, so it is fine for me :) But freenode would also work.

**Comment by javiercr** [comment]
@mikz, raul created an Slack channel, click on the Slack badge in the Readme and let's talk about this. :D
